### PR TITLE
Refactor Some Command-Related Methods in `aws_ssm.py`

### PR DESCRIPTION
**Pull request by beeankha**
##### SUMMARY

This work is part of the currently ongoing [AWS SSM Connection Refactoring & Plugin Promotion](https://forum.ansible.com/t/aws-ssm-connection-refactoring-plugin-promotion/39930) effort and related to https://github.com/ansible-collections/amazon.aws/issues/2394.

Fixes [ACA-2095](https://issues.redhat.com/browse/ACA-2095)

##### ISSUE TYPE

- Feature Pull Request

##### TASK LIST

- [x] The [`_exec_transport_commands`](https://github.com/ansible-collections/community.aws/blob/f2bd35b1e5d9c333ae7f1d63f52ad97b6ad12466/plugins/connection/aws_ssm.py#L1011) method is refactored to accept a single structured object (a tuple containing lists of command dictionaries) that contains the necessary command data.
- [ ] ~~Create a custom dataclass to hold command output.~~
- [x] Create a ~~custom dataclass~~ `TypedDict` to hold command results.
- [x] The [`_generate_commands`](https://github.com/ansible-collections/community.aws/blob/7f0f9e0ebaf6a068fcad320bc41f3e45e952a84f/plugins/connection/aws_ssm.py#L956) function returns a list of typed dictionaries with clear metadata (command strings, method, headers).
- [x] [`_exec_transport_commands`](https://github.com/ansible-collections/community.aws/blob/7f0f9e0ebaf6a068fcad320bc41f3e45e952a84f/plugins/connection/aws_ssm.py#L1010) accepts a list of typed dictionaries as an argument and returns the custom `CommandResults` object.
- [x] Python type hints are added to the method signature and the structured object to improve readability, facilitate static analysis, and clearly define the expected structure and data types.
- [x] Docstrings should be added to any refactored methods and the structured object (if custom), clarifying the purpose, structure, inputs, outputs, and any edge cases or special handling (if applicable).
- [x] Unit tests have been created to ensure that the refactored [`_generate_commands`](https://github.com/ansible-collections/community.aws/blob/7f0f9e0ebaf6a068fcad320bc41f3e45e952a84f/plugins/connection/aws_ssm.py#L956) method outputs the expected structured command object.
- [x] Add a changelog file.

### CODE REVIEW

File: tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py [hunk 1]
```diff
@@ -268,3 +269,72 @@ def test_generate_mark(self):
...
+        assert test_command_generation[0][2]["method"] == "put"
+
+    @patch("ansible_collections.community.aws.plugins.connection.aws_ssm.Connection.is_windows")
+    def test_generate_commands_windows(self, mock_is_windows):
+        """Testing command generation on Windows systems"""
+        pc = PlayContext()
+        new_stdin = StringIO()
+        conn = connection_loader.get("community.aws.aws_ssm", pc, new_stdin)
+
+        mock_is_windows.return_value = True
```
Comment (abikouo) [inline tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py:314]:
```suggestion
   @pytest.mark.parametrize("is_windows", [True, False])
   @patch("ansible_collections.community.aws.plugins.connection.aws_ssm.Connection.is_windows")
    def test_generate_commands_windows(self, mock_is_windows, is_windows):
        """Testing command generation on Windows systems"""
        pc = PlayContext()
        new_stdin = StringIO()
        conn = connection_loader.get("community.aws.aws_ssm", pc, new_stdin)

        mock_is_windows.return_value = is_windows
```

File: tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py [hunk 2]
```diff
@@ -268,3 +269,72 @@ def test_generate_mark(self):
...
+
+        # Two command dictionaries are generated for Windows
+        assert len(test_command_generation[0]) == 2
+
+        # Check contents of command dictionaries
+        assert "command" in test_command_generation[0][0]
+        assert "method" in test_command_generation[0][1]
+        assert "headers" in test_command_generation[0][1]
+        assert "Invoke-WebRequest" in test_command_generation[0][1]["command"]
+        assert test_command_generation[0][1]["method"] == "put"
```
Comment (abikouo) [inline tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py:340]:
```suggestion
        if not is_windows:
          # Check contents of command dictionaries
          assert "command" in test_command_generation[0][0]
          assert "method" in test_command_generation[0][1]
          assert "headers" in test_command_generation[0][1]
          assert "Invoke-WebRequest" in test_command_generation[0][1]["command"]
          assert test_command_generation[0][1]["method"] == "put"
       else:
         # Check contents of command dictionaries
         assert "command" in test_command_generation[0][0]
         assert "method" in test_command_generation[0][2]
         assert "headers" in test_command_generation[0][2]
         assert "curl --request PUT -H" in test_command_generation[0][2]["command"]
         assert test_command_generation[0][2]["method"] == "put"
```

**Review by abikouo — CHANGES_REQUESTED** [review]
There is no need to add a test skipped depending on the platform because the `is_windows` param is computed from the EC2 instance, not the source computer.
You can combine both tests into a single one.

**Review by abikouo — APPROVED** [review]
LGTM!!

**Review by brahmanim — APPROVED** [review]
LGTM

File: plugins/connection/aws_ssm.py [hunk 3]
```diff
@@ -433,6 +436,17 @@ def filter_ansi(line: str, is_windows: bool) -> str:
     return line
 
 
+@dataclass
```
Comment (alinabuzachis) [inline plugins/connection/aws_ssm.py:439]: I would have used a typed dict instead of a dataclass, but it's ok.

**Review by alinabuzachis — APPROVED** [review]

**Review by GomathiselviS — APPROVED** [review]

File: plugins/connection/aws_ssm.py (hunk 3 above)
Comment (tremble) [reply plugins/connection/aws_ssm.py:439]: We've seen both starting to appear, is this something we want to be opinionated about?

File: tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py [hunk 4]
```diff
@@ -268,3 +268,46 @@ def test_generate_mark(self):
...
+            assert test_command_generation[0][1]["method"] == "put"
+            # Two command dictionaries are generated for Windows
+            assert len(test_command_generation[0]) == 2
+        else:
+            assert "method" in test_command_generation[0][2]
+            assert "headers" in test_command_generation[0][2]
+            assert "curl --request PUT -H" in test_command_generation[0][2]["command"]
+            assert test_command_generation[0][2]["method"] == "put"
+            # Three command dictionaries are generated on non-Windows systems
+            assert len(test_command_generation[0]) == 3
```
Comment (gravesm) [inline tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py:308]: Not a blocking suggestion, but it seems like we ought to be able to do a more complete test against the contents of `test_command_generation`. Since you're mocking `generate_presigned_url()` I think the return value for `_generate_commands()` shouldn't vary.

Comment (beeankha) [reply tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py:308]: I might be misunderstanding, but the command dictionaries that get generated contain different commands depending on whether the instances are [Windows](https://github.com/ansible-collections/community.aws/pull/2248/files?diff=split&w=0#diff-8000d7fb9262e11565b61882567d729fe0958cd9b1a0105683d5bbe0e5b4e585R995) vs [non-Windows](https://github.com/ansible-collections/community.aws/pull/2248/files?diff=split&w=0#diff-8000d7fb9262e11565b61882567d729fe0958cd9b1a0105683d5bbe0e5b4e585R1022), so I wanted to make sure that each variation of the commands were output as expected. The details like the URL and S3 path are the same; I do see some assertions I could pull out of the `if/else` block, though, so I'll update the test to reflect that.

Comment (gravesm) [reply tests/unit/plugins/connection/aws_ssm/test_aws_ssm.py]: Yeah, sorry, let me clarify. The tests currently test the structure of what's returned from `_generate_command()` and partially test the content. What I'm suggesting is that I think we should be able to test the full command that gets returned, rather than just testing that some portion of it contains a specific string. I would argue that if something were to change the command, it should break the test.

If I'm reading the code wrong and it's going to be difficult to test the full command, I think it's fine as is.
